pennylane/ops/arithmetic/adjoint.py

```diff
@@ -0,0 +1,435 @@
+        ``T`` are natively supported by ``default.qubit``. Other gates will be expanded either using a custom
+        adjoint decomposition defined in :meth:`~.operation.Operator.adjoint`.
+
+        .. code-block:: python3
+
+            def subroutine(wire):
+                qml.RX(0.123, wires=wire)
+                qml.RY(0.456, wires=wire)
+
+            dev = qml.device('default.qubit', wires=1)
+            @qml.qnode(dev)
+            def circuit():
+                subroutine(0)
+                qml.adjoint(subroutine)(0)
+                return qml.expval(qml.PauliZ(0))
+
+        This creates the following circuit:
+
+        >>> print(qml.draw(circuit)())
+        0: ──RX(0.12)──S──S†──RX(0.12)†─┤  <Z>
+        >>> print(qml.draw(circuit, expansion_strategy="device")())
+        0: ──RX(0.12)──S──S†──RX(-0.12)─┤  <Z>
+
+        **Single operation**
+
+        You can also easily adjoint a single operation just by wrapping it with ``adjoint``:
+
+        .. code-block:: python3
+
+            dev = qml.device('default.qubit', wires=1)
+            @qml.qnode(dev)
+            def circuit():
+                qml.RX(0.123, wires=0)
+                qml.adjoint(qml.RX)(0.123, wires=0)
+                return qml.expval(qml.PauliZ(0))
+
+        This creates the following circuit:
+
+        >>> print(qml.draw(circuit)())
+        0: ──RX(0.12)──RX(0.12)†─┤  <Z>
+
+
+        :title: Developer details
+
+        **Lazy Evaluation**
+
+        When ``lazy=False``, the function first attempts operation-specific decomposition of the
+        adjoint via the :meth:`.operation.Operator.adjoint` method. Only if an Operator doesn't have
+        an :meth:`.operation.Operator.adjoint` method is the object wrapped with the :class:`~.ops.arithmetic.Adjoint`
+        wrapper class.
+
+        >>> qml.adjoint(qml.PauliZ, lazy=False)(0)
+        PauliZ(wires=[0])
+        >>> qml.adjoint(qml.RX, lazy=False)(1.0, wires=0)
+        RX(-1.0, wires=[0])
+        >>> qml.adjoint(qml.S, lazy=False)(0)
+        Adjoint(S)(wires=[0])
+
+    """
+    if not callable(fn):
+        raise ValueError(
+            f"The object {fn} of type {type(fn)} is not callable. "
+            "This error might occur if you apply adjoint to a list "
+            "of operations instead of a function or template."
+        )
+
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        with stop_recording(), QuantumTape() as tape:
+            fn(*args, **kwargs)
+
+        if lazy:
+            adjoint_ops = [Adjoint(op) for op in reversed(tape.operations)]
+            return adjoint_ops[0] if len(adjoint_ops) == 1 else adjoint_ops
+
+        def _op_adjoint(op):
+            try:
+                return op.adjoint()
+            except AdjointUndefinedError:
+                return Adjoint(op)
+
+        adjoint_ops = [_op_adjoint(op) for op in reversed(tape.operations)]
+        return adjoint_ops[0] if len(adjoint_ops) == 1 else adjoint_ops
```

Comment:
Just style wise, you could probably add an else block to get rid of the redundant return line
```suggestion
    def wrapper(*args, **kwargs):
        with stop_recording(), QuantumTape() as tape:
            fn(*args, **kwargs)

        if lazy:
            adjoint_ops = [Adjoint(op) for op in reversed(tape.operations)]

        else:
            def _op_adjoint(op):
                try:
                    return op.adjoint()
                except AdjointUndefinedError:
                    return Adjoint(op)

            adjoint_ops = [_op_adjoint(op) for op in reversed(tape.operations)]
        return adjoint_ops[0] if len(adjoint_ops) == 1 else adjoint_ops
```

Stick with which ever one you prefer

Reply:
done.